File: backend/internal/social/like/entity/like.go
// DOMAIN: SOCIAL
// NOTE: Content engagement through likes

package entity

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// TargetType represents the type of entity that can be liked.
//
// - content: All content items (posts and requests)
// - comment: Comments on content
type TargetType string

const (
	TargetTypeContent TargetType = "content"
	TargetTypeComment TargetType = "comment"
)

// Like represents a like on content by a user.
// Likes are immutable - once created, they cannot be modified.
// To remove a like, delete the record.
type Like struct {
	ContentID uuid.UUID
	UserID    uuid.UUID
	CreatedAt time.Time
}

// ErrTargetNotFound is returned when target doesn't exist.
type ErrTargetNotFound struct {
	TargetID   uuid.UUID
	TargetType TargetType
}

func (e *ErrTargetNotFound) Error() string {
	return fmt.Sprintf("target not found: target_type=%s, target_id=%s", e.TargetType, e.TargetID)
}

// ErrContentNotFound is returned when content doesn't exist.
type ErrContentNotFound struct {
	ContentID uuid.UUID
}

func (e *ErrContentNotFound) Error() string {
	return fmt.Sprintf("content not found: content_id=%s", e.ContentID)
}

// ErrContentDeleted is returned when trying to like deleted content.
type ErrContentDeleted struct {
	ContentID uuid.UUID
}

func (e *ErrContentDeleted) Error() string {
	return fmt.Sprintf("content is deleted: content_id=%s", e.ContentID)
}

// ErrInvalidTargetType is returned when an invalid target type is provided.
type ErrInvalidTargetType struct {
	TargetType TargetType
}

func (e *ErrInvalidTargetType) Error() string {
	return fmt.Sprintf("invalid target type: %s", e.TargetType)
}

// ErrLikeStatsInaccessible is returned when like stats are requested for a
// target that is not visible to the viewer (deleted/hidden content, blocked
// relationship, or nonexistent target). It lets the handler answer 404 without
// leaking engagement metadata.
var ErrLikeStatsInaccessible = fmt.Errorf("like stats not accessible")

// IsValidTargetType checks if a target type is valid for likes.
// Valid types: content, comment
func IsValidTargetType(t TargetType) bool {
	switch t {
	case TargetTypeContent, TargetTypeComment:
		return true
	default:
		return false
	}
}

File: apps/mobile/lib/domains/social/like/domain/entities/like.dart
/// Target types that can be liked
///
/// BACKEND ALIGNMENT V1:
/// - content: All content items
/// - comment: Comments on content
///
/// Aligned with backend: backend/internal/social/like/entity/like.go
/// Backend validates: oneof=content comment
enum LikeTargetType { content, comment }

/// Like statistics for a specific target
class LikeStats {
  final String targetId;
  final LikeTargetType targetType;
  final int totalLikes;
  final bool isLikedByCurrentUser;

  const LikeStats({
    required this.targetId,
    required this.targetType,
    required this.totalLikes,
    required this.isLikedByCurrentUser,
  });

  LikeStats copyWith({
    String? targetId,
    LikeTargetType? targetType,
    int? totalLikes,
    bool? isLikedByCurrentUser,
  }) {
    return LikeStats(
      targetId: targetId ?? this.targetId,
      targetType: targetType ?? this.targetType,
      totalLikes: totalLikes ?? this.totalLikes,
      isLikedByCurrentUser: isLikedByCurrentUser ?? this.isLikedByCurrentUser,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is LikeStats &&
        other.targetId == targetId &&
        other.targetType == targetType &&
        other.totalLikes == totalLikes &&
        other.isLikedByCurrentUser == isLikedByCurrentUser;
  }

  @override
  int get hashCode {
    return targetId.hashCode ^
        targetType.hashCode ^
        totalLikes.hashCode ^
        isLikedByCurrentUser.hashCode;
  }

  @override
  String toString() {
    return 'LikeStats(targetId: $targetId, targetType: $targetType, totalLikes: $totalLikes, isLikedByCurrentUser: $isLikedByCurrentUser)';
  }
}
